ICSharpCode.Decompiler/Disassembler/Adapters.cs

```diff
@@ -0,0 +1,218 @@
+﻿// Copyright (c) 2022 tom-englert.de for the SharpDevelop Team
+// 
+// Permission is hereby granted, free of charge, to any person obtaining a copy of this
+// software and associated documentation files (the "Software"), to deal in the Software
+// without restriction, including without limitation the rights to use, copy, modify, merge,
+// publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons
+// to whom the Software is furnished to do so, subject to the following conditions:
+// 
+// The above copyright notice and this permission notice shall be included in all copies or
+// substantial portions of the Software.
+// 
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+// INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
+// PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE
+// FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+// OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+#nullable enable
+
+using System;
+
+using ICSharpCode.Decompiler.Metadata;
+
+using System.Reflection.Metadata;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace ICSharpCode.Decompiler.Disassembler
+{
+	public abstract class Adapter
```

Comment:
I feel that a whole class hierarchy is a bit over-engineered for something that could just be two parameters, `MetadataReader` and `EntityHandle`.

Previously, when doing the big "SRM rewrite" of the decompiler engine, I added a lot of wrappers for all the different handle kinds, however almost all of them have disappeared because they do not provide much value over directly working with `MetadataReader` and a handle.

Would it be possible for you to describe your use-case for these classes? If it's about performance or allocations, I'd like to point out that we're calling `MetadataReader.GetString` all the time and it has not been a bottleneck so far. If you did some profiling, would it be possible for you to share the results? Thanks!

Reply:
I've added these adapters to make the exposed new interface as lean as possible; using an interface should not require to understand all the implementation details if not needed.
The adapters are as lean as possible, and I tried to touch the existing code as few as possible.

Every other approach ended up in a clumsy interface and a very complex implementation.

Performance: For methods it's not only `MetadataReader.GetString`, and using this in sorting will call it up to nearly N² times for N methods,  so ignoring this and waiting for the community to report performance problems sounds dangerous to me.

Reply:
You're adding 11 new classes to the public API, just for sorting a few lists. I think that's an absolute no-go.
If you think this is truly necessary, you will need to come up with some convincing arguments. If it's about perf, include actual benchmarks.

Reply:
Sorting isn't quadratic. I strongly suspect that when you run your benchmark you will find that the sorting has no influence on the overall disassembly performance.

Reply:
e.g. QSort does up to O(N²) comparisons, so you don't want to calculate the two comparands every time.

Reply:
But .NET's `Sort` isn't a stupid quadratic sort.
With any decent sort, there will be O(N lg N) string comparisons. Whether we perform O(N) or O(N lg N) `GetString` calls doesn't change the overall runtime -- that remains O(N lg N).
You're only improving performance by a small constant factor. But if the sorting algorithm only takes a negligible portion of the disassembly time to begin with, your performance improvement won't be big enough to be measurable.

Reply:
@dgrunwald The EnumerableSorter reads the keys only once, that's what makes the difference.